schema.go

```diff
@@ -579,6 +652,22 @@ func (s *RecordSchema) FingerprintUsing(typ FingerprintType) ([]byte, error) {
 	return s.fingerprinter.FingerprintUsing(typ, s)
 }
 
+// CacheFingerprint returns a special fingerprint of the schema for caching purposes.
+func (s *RecordSchema) CacheFingerprint() [32]byte {
+	data := make([]any, 0)
+	for _, field := range s.fields {
+		if field.HasDefault() {
```

Comment:
I made the record's `cacheFingerprint` logic more generic when I initially wrote it. I considered it much safer to incorporate fields' defaults as part of the decoder's cache default behavior for the records.

This goes against YAGNI; I should limit it to the case of composite records (especially ones with `FieldSetDefault`):

```go
func (s *RecordSchema) CacheFingerprint() [32]byte {
	data := make([]any, 0)
	for _, field := range s.fields {
		if field.HasDefault() && field.action == FieldSetDefault {
			data = append(data, field.Name(), field.Default())
		}
	}
	if len(data) == 0 {
		return s.Fingerprint()
	}
	data = append(data, s.Fingerprint())
	return s.cacheFingerprinter.fingerprint(data)
}
```

Contrary to `defaultDecoder,` `skipDecoder` is stateless, so the `FieldIgnore` action doesn't affect the state of the record Decoder and can be ignored in this logic.

Reply:
What still intrigues me here is the naming of `cacheFingerprint` is not specific to schema evolution or composite records. It can easily be used somewhere else later on.

Reply:
> Contrary to defaultDecoder, skipDecoder is stateless, so the FieldIgnore action doesn't affect the state of the record Decoder and can be ignored in this logic.

I dont think this is true, as it is not done at runtime but at planning time, so the resultant decode will always skip this field.

Reply:
You are right! I made an error; I added `FieldIgnore` action to the fingerprint:

```go
if field.action == FieldIgnore {
	data = append(data, field.Name()+string(FieldIgnore))
}
```